proposals/XXXX-recursive-protocol-constraints.md

```diff
@@ -0,0 +1,265 @@
+
+### `BidirectionalCollection`
+
+* Remove conformance to `_BidirectionalIndexable`
+* Change associated type: `associatedtype SubSequence : BidirectionalCollection`
+* Change associated type: `associatedtype Indices : BidirectionalCollection`
+
+### `Collection`
+
+* Remove conformance to `_Indexable`
+* Change associated type: `associatedtype SubSequence : Collection where SubSequence.Index == Index, SubSequence.Indices == Indices`
+* Change associated type: `associatedtype Indices : Collection where Indices.Iterator.Element == Index, Indices.Index == Index, Indices.SubSequence == Indices`
+
+### `Default*Indices` (all variants)
+
+* Declarations changed to `public struct Default*Indices<Elements : *Collection> : *Collection`
+
+### `IndexingIterator`
+
+* Declaration changed to `public struct IndexingIterator<Elements : Collection> : IteratorProtocol, Sequence`
+
+### `LazyFilter*Collection` (all variants)
+
+* Add default associated type conformance: `typealias SubSequence = ${Self}<Base.SubSequence>`
+
+### `LazyMap*Collection` (all variants)
+
+* Add default associated type conformance: `typealias SubSequence = ${Self}<Base.SubSequence>`
+
+### `Mirror`
+
+* Remove all `C.Subsequence` and `C.Indices` constraints from `init<Subject, C : Collection>(_ subject: Subject, children: C, displayStyle: DisplayStyle?, ancestorRepresentation: AncestorRepresentation)`
+* Remove all `C.Subsequence` and `C.Indices` constraints from `init<Subject, C : Collection>(_ subject: Subject, unlabeledChildren: C, displayStyle: DisplayStyle?, ancestorRepresentation: AncestorRepresentation)`
+
+### `MutableCollection`
+
+* Change associated type: `associatedtype SubSequence : MutableCollection`
+
+### `RandomAccessCollection`
+
+* Change associated type: `associatedtype SubSequence : RandomAccessCollection`
+* Change associated type: `associatedtype Indices : RandomAccessCollection`
+
+### `RangeReplaceableCollection`
+
+* Change associated type: `associatedtype SubSequence : RangeReplaceableCollection`
+
+### `Sequence`
+
+* Change associated type: `associatedtype SubSequence : Sequence where Iterator.Element == SubSequence.Iterator.Element, SubSequence.SubSequence == SubSequence`
+
+### `*Slice` (all variants)
+
+* Add default associated type conformance: `typealias Indices = Base.Indices`
+
+## Source compatibility
+
+From a source compatibility perspective, this is a purely additive change if the user's code is correctly written. It
+is possible that users may have written code which defines semantically incorrect associated types, which the compiler
+now rejects because of the additional constraints. We do not consider this scenario "source-breaking".
+
+The following source listing is an example of code accepted by the Swift 3.0.1 compiler which is nevertheless
+semantically incorrect, and would be rejected by a Swift compiler implementing this change.
+
+```swift
+struct BadSequence : Sequence, IteratorProtocol {
```

Comment:
I'll improve the example. I agree it's pretty stupid; I couldn't think of anything better.